package.json

```diff
@@ -16,12 +16,17 @@
   },
   "devDependencies": {
     "@babel/preset-typescript": "^7.16.7",
+    "@parcel/optimizer-data-url": "2.7.0",
+    "@parcel/transformer-inline-string": "^2.8.2",
     "@typescript-eslint/parser": "^5.13.0",
+    "assert": "^2.0.0",
     "eslint": "^8.14.0",
     "eslint-config-react-app": "^7.0.1",
     "eslint-plugin-prettier": "^4.0.0",
+    "events": "^3.3.0",
     "jest": "^27.5.1",
-    "prettier": "^2.3.2"
+    "prettier": "^2.3.2",
+    "stream-browserify": "^3.0.0"
```

Comment:
Where did all of these changes come from?

Reply:
If there is some import to a Node builtin modul (e.g. `stream`), Parcel autoinstalls the corresponding polyfill.

Reply:
(And same goes for the two Parcel plugins listed in here, which are installed on demand. And in this case, you forgot the bump them when upgrading to `"parcel": "^2.8.2"`)

Reply:
Oh awesome! I assumed it was something like this, thanks for commenting.

BTW, shipping this is currently blocked on some significant [size regressions](https://github.com/captbaritone/webamp/pull/1186#issuecomment-1261664731). Webamp lazy is the simpler case to debug since it doesn't pull in some of the heavy dependencies like JSZip. 

Using the bundle analyzer reporter gives a nice looking report, but it's dominated by "code from unknown sources". So my next step is to try to figure out what's contributing to this regression.

https://enchanting-khapse-0a90c1.netlify.app/webamp-lazy.html

If you happen to have any tips on how to diagnose, that'd be super helpful!

Reply:
`"isLibrary": true` makes the default `"optimize": false`, because you usually don't want published libraries to be minified. But you can override that of course (I assume you want minification as they're called `*.min.js`?):
```diff
--- packages/webamp/package.json
+++ packages/webamp/package.json
@@ -15,7 +15,8 @@
     "webamp": {
       "source": "js/webamp.js",
       "outputFormat": "esmodule",
-      "isLibrary": true
+      "isLibrary": true,
+      "optimize": true
     },
     "webamp-browser": {
       "source": "js/webampBrowser.js",
@@ -24,7 +25,8 @@
     "webamp-lazy": {
       "source": "js/webampLazy.tsx",
       "outputFormat": "esmodule",
-      "isLibrary": true
+      "isLibrary": true,
+      "optimize": true
     },
     "demo-site": {
       "source": "demo/index.html"
@@ -36,7 +38,7 @@
```

Reply:
Awesome. That has helped considerably! Still regressing by ~40%. It looks like the report claims "parcel-runtime" is taking up ~300kb, so that will be my next thing to dig into.

<img width="1661" alt="Screenshot 2023-01-08 at 1 52 56 PM" src="https://user-images.githubusercontent.com/162735/211220951-7cc70167-9755-4641-a93d-72b0faaa3691.png">


Things I can think of to investigate next:

1. Ensure I have comparable browser support (avoid unnecessary polyfill bloat)
2. Reproduce/validate that we have the same image compression. Webamp includes the original skin as a CSS file with embedded png images. The webpack setup had a special transform to ensure these images got optimized as part of the build process.

Reply:
That is definitely not representing the actual bundle correctly. Probably something slightly off with the sourcemaps.
Without minification, I get a big 1mb blob of unknown sources. And then running terser probably caused that source mapping to point to the runtime instead.

https://github.com/parcel-bundler/parcel/pull/8756 improves the sourcemaps due because there was some Windows CR LF in node_modules. That PR will be released on the nightly npm tag in ~3 hours (https://github.com/parcel-bundler/parcel/actions/workflows/nightly-release.yml). But even with that, the result isn't more helpful than before

Reply:
Hang on, it really is the runtime. Because it's the inline CSS... with huge inline pngs...

![Bildschirm­foto 2023-01-08 um 23 26 46](https://user-images.githubusercontent.com/4586894/211222234-aa3533d3-3d5c-41ce-a78f-e76b9ab6ffb6.png)

Reply:
Interesting. Thanks again for having a look. In order to distribute Webamp as a single js file that people can embed via a `<script>` tag, we inline the default skin as a CSS file with embedded pngs. It makes sense that this would be an uncommon case. So it sounds like there's a miss-attribution in the source maps that's attributing the CSS/pngs to the parcel runtime?

These CSS files are the same in the webpack build though, so there's still some difference that's causing the Parcel build to be larger, but knowing about this misattribution will at least help me avoid going down a blind alley.

Reply:
I would first try to verify that these inlined CSS string(s) are equally large in both cases (so including the PNGs base64 strings).

Reply:
Okay. It looks like the remaining difference for the lazy version is entirely attributable to the fact that I haven't yet ported my [image optimization](https://github.com/captbaritone/webamp/blob/master/packages/webamp/scripts/postcss-optimize-data-uri-pngs.js) to Parcel.

I tried manually checking the optimized version of the CSS into the repo and we are now at ~3% reduction in bundle size!

Reading the Parcel docs it implies that png images are automatically optimized. It looks like that does not apply to png images that exist as data uris in CSS files (that's what my custom Webpack loader does). Is that an optimization that Parcel could/should apply? It seems in keeping with Parcel's approach of automatically applying optimizations where it can, but I also get that this is quite an edge case.

Reply:
🎉 

>  Is that an optimization that Parcel could/should apply?  I also get that this is quite an edge case.

Yeah, but it should be possible to support in Parcel.

An alternative here (and this might even be preferable just from a code authoring perspective) is putting these inline PNGs into separate files and then using Parcel's feature to inline it as a data-url: https://parceljs.org/features/bundle-inlining/#inlining-as-a-data-url

Reply:
Yeah, I think a "normal" app would work this way, but Webamp has built in support for loading custom "skins", which means it converts the zip file into CSS (with embedded data uris) in the browser. The way I generate the default skin CSS file is I load the custom skin and copy out the generated CSS. This means there's not a reasonable way to get those (probably hundreds) of png files as actual files on the file system.

Reply:
As you saw, I raised the issue on the Parcel repo: https://github.com/parcel-bundler/parcel/issues/8758#issuecomment-1375637091

I'll probably just keep my existing setup unless that feature looks appealing to someone to implement.

Reply:
I'll see if I can fix that parcel issue you mentioned -- I'll recomment here if I get to it.